README
this is kuvert, a wrapper around sendmail or other MTAs that
does gpg signing/signing+encrypting transparently, based
on the content of your public keyring(s) and your preferences.

how it works:
-------------

you need to configure your MUA to submit mails to kuvert instead of 
directly. you configure kuvert either to present an SMTP server to
your MUA, or you make your MUA to use kuvert_submit instead of executing
/usr/sbin/sendmail. kuvert_submit will spool the mail
in kuvert's queue iff there is a suitable configuration file.

kuvert is the tool that takes care of mangling the email. it reads the 
queue periodically and handles emails in the queue: signing or encrypting
the mail, then handing it over to /usr/lib/sendmail or an external SMTP 
server for transport.

(why a queue? because i thought it might be useful to make sure that none of
your emails leaves your system without kuvert handing it. you might be 
very paranoid, and kill kuvert whenever you leave your box (and remove
the keyrings as well).)

installation:
-------------

on debian systems you simply install the kuvert package, construct
a suitable .kuvert configuration file and off you go. 
an example config file is provided 
at /usr/share/doc/kuvert/examples/dot-kuvert.

on other systems you need to do the following:

you need perl perl 5.004+, gpg and a raft of perl modules:
MIME::Parser, Mail::Address, Net::SMTPS, Sys::Hostname, Net::Server::Mail,
Authen::SASL, IO::Socket::INET, Filehandle, File::Slurp, File::Temp, Fcntl 
and Time::HiRes.
some of those are part of a standard perl intall, others you'll have to
get from your nearest CPAN archive and install.
optional: get linux-kernel keyutils package, the gpg-agent or some 
other passphrase cache of your choice.

run make, make install DESTDIR=/ as root
-> kuvert, kuvert_submit, the manpages and one helper module 
will be installed in /usr/bin, /usr/share/man/man1 and 
/usr/share/perl5/Net/Server/Mail/ESMTP/, respectively.

configuration:
--------------

read the manpages for kuvert(1) and kuvert_submit(1) and 
consult the example config file "dot-kuvert". you will need
to create your own config file as ~/.kuvert. sorry, no autoconfig here:
this step is too crucial for a mere robot to perform. 

then start kuvert and inject a testmail, look at the logs to check
if everything works correctly.

(historical note: kuvert came into existence in 1996 as pgpmail and
was used only privately until 99, when it was extended and renamed
to guard. some of my friends started using this software, and in 
2001 it was finally re-christened kuvert, extended even further
and debianized. in 2008 it received a major overhaul to also provide 
inbound smtp as submission mechanism, outbound smtp transport and better
controllability via email addresses. until 2008 kuvert supported pgp2.x.)

please report bugs to me, Alexander Zangerl, <az@snafu.priv.at>.

The original source can always be found at:
	http://www.snafu.priv.at/kuvert/

Copyright (C) 1999-2013 Alexander Zangerl

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License version 2
  as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License with
  the Debian GNU/Linux distribution in file /usr/share/common-licenses/GPL;
  if not, write to the Free Software Foundation, Inc., 59 Temple Place,
  Suite 330, Boston, MA  02111-1307  USA






debian/NEWS
kuvert (2.0.0) unstable; urgency=low

  kuvert has been completely revamped, with a number of new features
  and minus some legacy stuff.
  
  New: 
   full inbound and outbound support for SMTP
   support for gpg-agent
   simplified action settings
   overridable keys and actions in the comments of From: and To:
   simpler submission wrapper program

  Gone: 
   support for pgp2 (but RSA keys continue to work fine, with gpg)
   kuvert no longer stores passphrases itself, ever.
    
  Because of these changes the new config file format is different. 
  The new kuvert will not run with an old-style config file, but 
  does a rough auto-conversion on startup. It'll dump that skeleton
  in /tmp and tell you about it. You will need to go over that and 
  the kuvert manpage and adjust the config to your liking.

 -- Alexander Zangerl <az@debian.org>  Sun, 29 Jun 2008 21:32:14 +1000

kuvert (1.1.10) unstable; urgency=low

  The configuration options AGENTPATH and CLIENTPATH have been
  deprecated and support for a private q-agent process was dropped;
  while kuvert still suggests quintuple-agent, it is no longer favouring
  it and will work with any (reasonable) external passphrase store.
  
  kuvert will not start until you update your 
  personal .kuvert configuration file and remove AGENTPATH/CLIENTPATH
  and use the new GETSECRET and DELSECRET directives if you want
  to use an external passphrase store. 

 -- Alexander Zangerl <az@debian.org>  Fri,  4 Nov 2005 16:43:14 +1000



plainAUTH.pm
package Net::Server::Mail::ESMTP::plainAUTH;
use strict;
use base qw(Net::Server::Mail::ESMTP::Extension);
use MIME::Base64;

use vars qw( $VERSION );
$VERSION = '1.0';

# the following are required by nsme::extension 
# but not documented :(
sub init
{
    my ($self,$parent)=@_;
    $self->{AUTH}=();
    return $self;
}

# the smtp operations we add
sub verb
{
    return ( [ 'AUTH' => \&handle_auth, ],);
}

# what to add to the esmtp capabilities response
sub keyword
{
	return 'AUTH LOGIN PLAIN';
}

# what options to allow for mail from: auth
sub option
{
	return (['MAIL', 'AUTH' => sub { return; }]);
}

# and the actual auth handler
sub handle_auth
{
    my ($self,$args)=@_;
    my ($method,$param);
    $args=~/^(LOGIN|PLAIN)\s*(.*)$/ && (($method,$param)=($1,$2));

    if ($self->{AUTH}->{active})
    {
	delete $self->{AUTH}->{active};
	$self->reply(535, "Authentication phases mixed up.");
	return undef;		# if rv given, server shuts conn!
    }
    elsif ($self->{AUTH}->{completed})
    {
	$self->reply(504,"Already authenticated.");
	return undef;
    }
    elsif (!$method)
    {
	$self->reply(501,"Unknown authentication method.");
	return undef;
    }

    $self->{AUTH}->{active}=$method;
    
    if ($param eq '*') 
    {
	delete $self->{AUTH}->{active};
	$self->reply(501, "Authentication cancelled.");
	return undef;
    }

    if ($method eq 'PLAIN') 
    {
	if ($param)		# plain: immediate with args
	{
	    my (undef,$user,$pwd)=split(/\0/,decode_base64($param),3);
	    if (!$user)
	    {
		delete $self->{AUTH}->{active};
		$self->reply(535, "5.7.8 Authentication failed.");
		return undef;
	    }
	    return run_callback($self,$user,$pwd);
	}
	else			# plain: or empty challenge and then response
	{
	    $self->reply(334," ");
	    # undocumented but crucial: direct stuff to this method
	    $self->next_input_to(\&process_response);
	    return undef;
	}
    }
    elsif ($method eq 'LOGIN') 
    {
	# login is always two challenges
	$self->reply(334, "VXNlcm5hbWU6"); # username
	$self->next_input_to(\&process_response);
	return undef;
    }
}

# runs user-supplied callback on username and password
# responds success if callback succeeds
# sets complete if ok, clears active either way
sub run_callback
{
    my ($self,$user,$pass)=@_;
    my $ok;

    my $ref=$self->{callback}->{AUTH};
    if (ref $ref eq 'ARRAY' && ref $ref->[0] eq 'CODE') 
    {
	my $c=$ref->[0];
	$ok=&$c($self,$user,$pass);
    }
    if ($ok)
    {
	$self->reply(235, "Authentication successful");
	$self->{AUTH}->{completed}=1;
    }
    else
    {
	$self->reply(535,"Authentication failed.");
    }
    delete $self->{AUTH}->{active};
    return undef;
}

# deals with any response, based on active method
sub process_response
{
    my ($self,$args)=@_;

    if (!$self->{AUTH}->{active} || $self->{AUTH}->{completed})
    {
	delete $self->{AUTH}->{active};
	$self->reply(535, "Authentication phases mixed up.");
	return undef;
    }
    if (!$args)
    {
	delete $self->{AUTH}->{active};
	$self->reply(535, "5.7.8 Authentication failed.");
	return undef;
    }
    
    if ($self->{AUTH}->{active} eq "PLAIN")
    {
	# plain is easy: only one response containing everything
	my (undef,$user,$pwd)=split(/\0/,decode_base64($args),3);
	if (!$user)
	{
	    delete $self->{AUTH}->{active};
	    $self->reply(535, "5.7.8 Authentication failed.");
	    return undef;
	}
	return run_callback($self,$user,$pwd);
    }
    elsif ($self->{AUTH}->{active} eq "LOGIN")
    {
	# uglier: two challenges for username+password
	my ($input)=split(/\0/,decode_base64($args));

	# is this the second time round?
	if ($self->{AUTH}->{user})
	{
	    return run_callback($self,$self->{AUTH}->{user},$input);
	}
	else			
	{
	    # nope, first time: save username and challenge
	    # for password
	    $self->{AUTH}->{user}=$input;
	    $self->reply(334, "UGFzc3dvcmQ6"); # password
	    $self->next_input_to(\&process_response);
	    return undef;
	}
    }
    else 
    {
	delete $self->{AUTH}->{active};
	$self->reply(535, "Authentication mixed up.");
	return undef;
    }
}

1;


kuvert_submit.c
/*
 * 
 * this file is part of kuvert, a wrapper around your mta that
 * does pgp/gpg signing/signing+encrypting transparently, based
 * on the content of your public keyring(s) and your preferences.
 *
 * copyright (c) 1999-2008 Alexander Zangerl <az+kuvert@snafu.priv.at>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdio.h>
#include <pwd.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <ctype.h>
#include <syslog.h>
#include <stdlib.h>

#define CONFFILE "/.kuvert"
#define DEFAULT_QUEUEDIR "/.kuvert_queue"
#define BUFLEN 65536
#define FALLBACKMTA "/usr/sbin/sendmail"

#define BAILOUT(a,...) {fprintf(stderr,"%s: ",argv[0]); fprintf(stderr, a "\n",##__VA_ARGS__);syslog(LOG_ERR,a,##__VA_ARGS__); exit(1);}

int main(int argc,char **argv)
{
   struct passwd *pwentry;
   /* fixme sizes */
   char filen[256],buffer[BUFLEN],dirn[256];
   int res,c,spaceleft;
   char *p,*dirnp;
   FILE *out;
   FILE *cf;
   struct stat statbuf;
   int direct=1,norecips=0,testmode=0,i;

   /* determine whether to queue stuff or to call sendmail
      directly: if there is no proper config file for kuvert in $HOME
      or if given -bv go direct, otherwise we enqueue.  */
   openlog(argv[0],LOG_NDELAY|LOG_PID,LOG_MAIL);

   for(i=1;i<argc;++i)
   {
      if (!strncmp(argv[i],"-bv",3))
      {
	 testmode=1;
	 syslog(LOG_INFO,"-bv argument present, running sendmail.");
	 break;
      }
   }
   
   if (!testmode)
   {
      /* look for config file in $HOME */
      pwentry=getpwuid(getuid());
      if (!pwentry)
	 BAILOUT("getpwuid failed: %s",strerror(errno));
      
      /* open and scan the conffile for an queue-file definition 
	 if there is no conffile, kuvert wont work ever  */
      if (snprintf(filen,sizeof(filen),"%s%s",pwentry->pw_dir,CONFFILE)==-1)
	 BAILOUT("overlong filename, suspicious",NULL);
      if (!(cf=fopen(filen,"r")))	
      {
	 /* no config file -> exec sendmail */
	 syslog(LOG_INFO,"user has no "CONFFILE" config file, running sendmail");
      }
      else
      {
	 direct=0;
	 /* scan the lines for ^QUEUEDIR\s+ */
	 dirnp=NULL;
	 while(!feof(cf))
	 {
	    p=fgets(buffer,sizeof(buffer)-1,cf);
	    /* empty file? ok, we'll ignore it */
	    if (!p)
	       break;
	    
	    if (!strncasecmp(buffer,"QUEUEDIR",sizeof("QUEUEDIR")-1))
	    {
	       p=buffer+sizeof("QUEUEDIR")-1;
	       for(;*p && isspace(*p);++p)
		  ;
	       if (*p)
	       {
		  dirnp=p;
		  /* strip the newline from the string */
		  for(;*p && *p != '\n';++p)
		     ;
		  if (*p == '\n')
		     *p=0;
		  /* strip eventual trailing whitespace */
		  for(--p;p>dirnp && isspace(*p);--p)
		     *p=0;
	       }
	       /* empty dir? ignore it */
	       if (strlen(dirnp)<2)
		  dirnp=NULL;
	       break;
	    }
	 }
	 fclose(cf);
      }
   }

   /* direct to sendmail requested? */
   if (direct)
   {
      /* mangle argv[0], so that it gets recognizeable by sendmail */
      argv[0]=FALLBACKMTA;
      *buffer=0;

      /* bah, c stringhandling is ugly... i just want all args 
	 in one string for a nice syslog line... */
      for(c=0,spaceleft=sizeof(buffer);
	  c<argc;
	  spaceleft-=strlen(argv[c++]))
      {
	 if (spaceleft <= 0) 
	    BAILOUT("overlong command line, suspicious.",NULL);
	 strncat(buffer,argv[c],spaceleft);
	 --spaceleft && c<argc-1 && strcat(buffer," ");
      }
    
      syslog(LOG_INFO,"executing MTA '%s' directly",buffer);
      execv(FALLBACKMTA,argv);
      /* must not reach here */
      BAILOUT("execv "FALLBACKMTA" failed: %s",strerror(errno));
   }

   /* otherwise queue the stuff for kuvert,
      first check queuedir and create if missing */
   if (!dirnp)
   {
      if(snprintf(dirn,sizeof(dirn),"%s%s",pwentry->pw_dir,DEFAULT_QUEUEDIR)
	 ==-1)
	 BAILOUT("overlong dirname, suspicous.",NULL);
      dirnp=dirn;
   }

   res=stat(dirnp,&statbuf);
   if (res)
   {
      if (errno == ENOENT)
      {
	 /* seems to be missing -> try to create it */
	 if (mkdir(dirnp,0700))
	    BAILOUT("mkdir %s failed: %s\n",dirnp,strerror(errno));
      }
      else 
	 BAILOUT("stat %s failed: %s\n",dirnp,strerror(errno));
   }
   else if (!S_ISDIR(statbuf.st_mode))
   {  
      BAILOUT("%s is not a directory",dirnp);
   }
   else if (statbuf.st_uid != getuid())
   {
      BAILOUT("%s is not owned by you - refusing to run",dirnp);
   }
   else if ((statbuf.st_mode & 0777) != 0700)
   {
      BAILOUT("%s does not have mode 0700 - refusing to run",dirnp);
   }
   umask(066);			/* absolutely no access for group/others... */
   
   /* dir does exist now */
   snprintf(filen,sizeof(filen),"%s/%d",dirnp,getpid());
  
   /* file create and lock */
   if (!(out=fopen(filen,"a")))
   {
      BAILOUT("fopen %s failed: %s\n",filen,strerror(errno));
   }
   if (flock(fileno(out),LOCK_EX))
   {
      BAILOUT("flock failed: %s\n",strerror(errno));
   }

   /* scan the arguments for the LSB-mandated options:
      we ignore any options but -f, -t.
   /* no getopt error messages, please! */
   opterr=0;
   while ((c=getopt(argc,argv,"f:t"))!=-1)
   {
      if (c=='?')
	 continue;		/* we simply ignore uninteresting options */
      else if (c=='f')
      {
	 /* pass the intended envelope sender */
	 fprintf(out,"X-Kuvert-From: %s\n",optarg);
      }
      else if (c=='t')
      {
	 /* no recipients given, so we don't need to pass any recips */
	 norecips=1;
      }
   }

   if (!norecips && optind<argc)
   {
      fprintf(out,"X-Kuvert-To: ");
      for(c=optind;c<argc;++c)
      {
	 fprintf(out,"%s%s",argv[c],(c<argc-1?", ":"\n"));
      }
   }
   fflush(out);

   /* now finally put the data in the queuefile */
   do 
   {
      res=fread(buffer,1,BUFLEN,stdin);
      if (!res && ferror(stdin))
	 BAILOUT("fread failure: %s",strerror(errno));
      if (fwrite(buffer,1,res,out)!=res && ferror(out))
	 BAILOUT("fwrite failure: %s",strerror(errno));
   }
   while (res==BUFLEN);
  
   if (fflush(out)==EOF)
      BAILOUT("fflush failed: %s",strerror(errno));
   if (flock(fileno(out),LOCK_UN))
   {
      BAILOUT("flock (unlock) failed: %s",strerror(errno));
   }
   if (fclose(out)==EOF)
      BAILOUT("fclose failed: %s",strerror(errno));
   return 0;
}


kuvert_mta_wrapper.c
/*
 * $Id: kuvert_mta_wrapper.c,v 1.8 2007/06/23 03:14:46 az Exp az $
 * 
 * this file is part of kuvert, a wrapper around your mta that
 * does pgp/gpg signing/signing+encrypting transparently, based
 * on the content of your public keyring(s) and your preferences.
 *
 * copyright (c) 1999-2003 Alexander Zangerl <az+kuvert@snafu.priv.at>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdio.h>
#include <pwd.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <ctype.h>
#include <syslog.h>
#include <stdlib.h>

#define CONFFILE "/.kuvert"
#define DEFAULT_QUEUEDIR "/.kuvert_queue"
#define BUFLEN 65536
#define FALLBACKMTA "/usr/lib/sendmail"

#define BAILOUT(a,...) {fprintf(stderr,"%s: ",argv[0]); fprintf(stderr, a "\n",##__VA_ARGS__);syslog(LOG_ERR,a,##__VA_ARGS__); exit(1);}

int main(int argc,char **argv)
{
   struct passwd *pwentry;
   /* fixme sizes */
   char filen[256],buffer[BUFLEN],dirn[256];
   int res,c,fallback=0,spaceleft;
   char *p,*dirnp;
   FILE *out;
   FILE *cf;
   struct stat statbuf;

   /* determine whether to queue stuff or to call sendmail
      directly: if there is a proper config file of kuvert in $HOME,
      and if the flags/args given are "consistent" with a call
      to sendmail for mail submission, do queue stuff;
      otherwise exec sendmail. */

   openlog(argv[0],LOG_NDELAY|LOG_PID,LOG_MAIL);
   
   /* scan the arguments for options:
      we understand about: no options, non-option-args, --,
      -bm, -f, -i, -t, -v, -m, -oi, -d*, -e*. everything else means some special
      instruction to sendmail, so we exec sendmail. */

   /* no getopt error messages, please! */
   opterr=0;

   while ((c=getopt(argc,argv,"f:itvb:mo:"))!=-1 && !fallback)
   {
      switch (c)
      {
	 case 'v':
	 case 'f':
	 case 'i':
	 case 't':
	 case 'm':			/* deprecated option 'metoo',
					   but nmh uses this... */
	    break;			/* these options are ok and supported */
	 case 'b':
	    /* just -bm is ok, other -b* are bad */
	    if (!optarg || *optarg != 'm') 
	    {
	       fallback=1;
	       syslog(LOG_INFO,"option '-%c%s' mandates fallback",
		      c,optarg ? optarg : "");
	    }
	    break;
	 case 'o':
	    /* -oi, -oe*, -od* are ok */
	    if (!optarg || (*optarg != 'i' && *optarg != 'e' 
			    && *optarg != 'd'))
	    {
	       fallback=1;
	       syslog(LOG_INFO,"option '-%c%s' mandates fallback",
		      c,optarg ? optarg : "");
	    }
	    break;
	 default:
	    /* well, there's an option we do not know, lets bail out */
	    fallback=1;
	    syslog(LOG_INFO,"option '-%c' mandates fallback",
		   c=='?'?optopt:c);
	    break;
      }
   }
   if (!fallback)
   {
      /* options seem ok, look for config file in $HOME */
      pwentry=getpwuid(getuid());
      if (!pwentry)
	 BAILOUT("getpwuid failed: %s",strerror(errno));
    
      /* open and scan the conffile for an queue-file definition 
	 if there is no conffile, kuvert wont work ever  */
      if (snprintf(filen,sizeof(filen),"%s%s",pwentry->pw_dir,CONFFILE)==-1)
	 BAILOUT("overlong filename, suspicious",NULL);
      if (!(cf=fopen(filen,"r")))	
      {
	 /* no config file -> exec sendmail */
	 syslog(LOG_INFO,"user has no .kuvert config file, fallback");
	 fallback=1;
      }
      else
      {
	 /* scan the lines for ^QUEUEDIR\s+ */
	 dirnp=NULL;
	 while(!feof(cf))
	 {
	    p=fgets(buffer,sizeof(buffer)-1,cf);
	    /* empty file? ok, we'll ignore it */
	    if (!p)
	       break;
	
	    if (!strncmp(buffer,"QUEUEDIR",sizeof("QUEUEDIR")-1))
	    {
	       p=buffer+sizeof("QUEUEDIR")-1;
	       for(;*p && isspace(*p);++p)
		  ;
	       if (*p)
	       {
		  dirnp=p;
		  /* strip the newline from the string */
		  for(;*p && *p != '\n';++p)
		     ;
		  if (*p == '\n')
		     *p=0;
		  /* strip eventual trailing whitespace */
		  for(--p;p>dirnp && isspace(*p);--p)
		     *p=0;
	       }
	       /* empty dir? ignore it */
	       if (strlen(dirnp)<2)
		  dirnp=NULL;
	       break;
	    }
	 }
	 fclose(cf);
      }
   }

   /* fallback to sendmail requested? */
   if (fallback)
   {
      /* mangle argv[0], so that it gets recognizeable by sendmail */
      argv[0]=FALLBACKMTA;
      *buffer=0;

      /* bah, c stringhandling is ugly... i just want all args 
	 in one string for a nice syslog line... */
      for(c=0,spaceleft=sizeof(buffer);
	  c<argc;
	  spaceleft-=strlen(argv[c++]))
      {
	 if (spaceleft <= 0) 
	    BAILOUT("overlong command line, suspicious.",NULL);
	 strncat(buffer,argv[c],spaceleft);
	 --spaceleft && c<argc-1 && strcat(buffer," ");
      }
    
      syslog(LOG_INFO,"will exec MTA as '%s'",buffer);
      execv(FALLBACKMTA,argv);
      /* must not reach here */
      BAILOUT("execv FALLBACKMTA failed: %s",strerror(errno));
   }

   /* otherwise queue the stuff for kuvert,
      first check queuedir and create if missing */
   if (!dirnp)
   {
      if(snprintf(dirn,sizeof(dirn),"%s%s",pwentry->pw_dir,DEFAULT_QUEUEDIR)
	 ==-1)
	 BAILOUT("overlong dirname, suspicous.",NULL);
      dirnp=dirn;
   }

   res=stat(dirnp,&statbuf);
   if (res)
   {
      if (errno == ENOENT)
      {
	 /* seems to be missing -> try to create it */
	 if (mkdir(dirnp,0700))
	    BAILOUT("mkdir %s failed: %s\n",dirnp,strerror(errno));
      }
      else 
	 BAILOUT("stat %s failed: %s\n",dirnp,strerror(errno));
   }
   else if (!S_ISDIR(statbuf.st_mode))
   {  
      BAILOUT("%s is not a directory",dirnp);
   }
   else if (statbuf.st_uid != getuid())
   {
      BAILOUT("%s is not owned by you - refusing to run",dirnp);
   }
   else if ((statbuf.st_mode & 0777) != 0700)
   {
      BAILOUT("%s does not have mode 0700 - refusing to run",dirnp);
   }
   umask(066);			/* absolutely no access for group/others... */
    
   /* dir does exist now */
   snprintf(filen,sizeof(filen),"%s/%d",dirnp,getpid());
  
   /* file create and lock */
   if (!(out=fopen(filen,"a")))
   {
      BAILOUT("fopen %s failed: %s\n",filen,strerror(errno));
   }
   if (flock(fileno(out),LOCK_EX))
   {
      BAILOUT("flock failed: %s\n",strerror(errno));
   }

   /* and put the data there */
   do 
   {
      res=fread(buffer,1,BUFLEN,stdin);
      if (!res && ferror(stdin))
	 BAILOUT("fread failure: %s",strerror(errno));
      if (fwrite(buffer,1,res,out)!=res && ferror(out))
	 BAILOUT("fwrite failure: %s",strerror(errno));
   }
   while (res==BUFLEN);
  
   if (fflush(out)==EOF)
      BAILOUT("fflush failed: %s",strerror(errno));
   if (flock(fileno(out),LOCK_UN))
   {
      BAILOUT("flock (unlock) failed: %s",strerror(errno));
   }
   if (fclose(out)==EOF)
      BAILOUT("fclose failed: %s",strerror(errno));
   return 0;
}
